lib/Codegen/Lowering/ForEquationLowerer.cpp

```diff
@@ -67,8 +68,10 @@ namespace marco::codegen::lowering
       builder().setInsertionPointToStart(forEquationOp.bodyBlock());
 
       // Add the induction variable to the symbol table
+      const llvm::StringRef name = forIndex->getName();
+      getDeclaredVariables().insert(std::string(name));
```

Comment:
When is the variable going to be removed from the set?

Reply:
The variable is never removed from the set. "declaredVariables" contains every variable identifier that has ever been inserted into "variablesSymbolTable". When an error with a variable identifier occurs, there is a function to check, for each element of the set, whether it is visible from the location that generated the error or not.

Reply:
So, in general, the content of set is not coherent with the content of the symbol table. Let's make it coherent, as in future the set may be used for other reasons and the current state would lead to wrong behaviours.

Reply:
If I understood correctly, variablesSymbolTable has all of its elements removed when the destructor of LookupScopeGuard is invoked, and this is the only place where elements are removed from it. Therefore, I could simply modify the same destructor to remove all elements from declaredVariables as well. Is this correct?

Reply:
I wasn't referring to changing the `VariablesSymbolTable` destructor, but rather the `LookupScopeGuard` destructor. 
Since there is currently no way to iterate on the variables declared in a scope, this change would require storing the scope information in `declaredVariables` as well, basically reimplementing a `llvm::ScopedHashTable`.
For example, I could change `declaredVariables` into a map with scopes as keys, and `std::set<std::string>` as values, then edit the `LookupScopeGuard` destructor to remove all entries that have a scope in `lookupScopes` as keys. Is this what I should do?

Reply:
Declared variables and lookup scopes are somewhat close to each other but I would keep them separate for now. You can just make the current `VariablesSymbolTable` a new class wrapping the current `llvm::ScopedHashTable` and the `set` you need, and then implement a scope class along the same lines. This way you don't have to reimplement anything but a proxy.

Reply:
I implemented the classes you suggested. 
However, the attribute `variablesSet` of `VariablesSymbolTable` still contains all variables, with no removals. The reason is that I realized that this information is needed elsewhere, since the same set is used in cases such as the one in `ParserError/fields/field_scope.mo`, to find the identifiers `fild` and `field`. In fact, while I can iterate on nested `mlir::SymbolTable`s, I cannot actually iterate on the variables declared in a single `mlir::SymbolTable`.
I believe the new code is still cleaner than the old one, as the only way to accidentally access variables that are not in the `llvm::ScopedHashTable` is using the new `getVariables` method of the `VariablesSymbolTable` with the input `onlyVisible = false`, and the set is now hidden inside the class.